### ISSUE TITLE
How to swap subviews?

### ISSUE THREAD

**Issue by Veid** [leeoniya/domvm#183]
How can I swap subviews in a view? For example, this doesn't work.

```js
var el = domvm.defineElement;
var vw = domvm.defineView;
var iv = domvm.injectView;

function ViewA(vm, args) {
    console.log('viewA init');

    return function() {
        return el('span', 'viewA');
    }
}

function ViewB(vm, args) {
    console.log('viewB init');

    return function() {
        return el('span', 'viewB');
    }
}

function TabView(vm, args) {
    return function() {
        return el('div', [
            el('ol', args.tabs.map(function(tab) {
                return el('li', tab.text);
            })),
            renderTabContent(args.selectedView)
        ]);
    }

    function renderTabContent(selectedView) {
        console.log(selectedView);

        if (!selectedView) {
            return undefined;
        }

        return el('div#tabContent', [ iv(selectedView) ]);
    }
}

var aVM = domvm.createView(ViewA);
var bVM = domvm.createView(ViewB);
var allViews = [aVM, bVM];

var args = {
    tabs         : [{text:'abc'}, {text:'def'}],
    selectedView : aVM
}

var tabVM = domvm.createView(TabView, args);
tabVM.mount(document.body);

var i = 0;
setInterval(function() {
    args.selectedView = allViews[i%2];
    tabVM.redraw();
    i += 1;
}, 1000);
```

 I can make it work by using `defineView` instead of `injectView` (and changing `selectedView` to be the view instead of the vm, etc.). However, that seems to always re-create the view, calling the init every time.

How can I define several views, then switch between them in a main view?

**Comment by leeoniya** [comment]
i copy-pasted your example and it appears to work okay?: https://jsfiddle.net/o6bzvcc4/

**Comment by Veid** [comment]
Odd. See this fiddle: https://jsfiddle.net/s6v2mf9b/

**Comment by leeoniya** [comment]
looks like the same bug i fixed a few days ago here: https://github.com/leeoniya/domvm/issues/182

:)

**Comment by Veid** [comment]
Ah brilliant, thank you! Do you know when you'll incorporate the fix into a new build?

**Comment by leeoniya** [comment]
probably when that issue is resolved. i think my final diagnosis there is correct, but i'll wait for @tropperstyle to confirm.

i'll likely add a devmode warning if an already-hydrated-but-yet-unmounted vm ever reaches the mounting phase in the dom reconciler's `syncChildren()`.

**Comment by leeoniya** [comment]
v3.2.3 with these patches is tagged.

**Comment by Veid** [comment]
Got it, thanks.

Moving on, is the above code even a good way to do this? What if my ViewA was something like this instead:

```js
function ViewA(vm, args) {
	var text = 'viewA';

	setInterval(function() {
		text = text.split('').sort(function(){return 0.5-Math.random()}).join('');
		vm.parent().redraw();
	}, 500);

	return function() {
		return el('span', text);
	}
}
```

This won't work because if `ViewB` is currently selected instead, then `ViewA`'s `vm.node` will be null.

I'm doing something similar to this in a different piece of code that's too large to paste here, but I've run into sporadically. So I'm wondering if this isn't the best way to go about doing thing.

```
UNMOUNTED_REDRAW: Invoking redraw() of an unmounted (sub)view may result in errors. 
```

Would it be better to actually render all the tabs and display/show them instead?

**Comment by leeoniya** [comment]
the original issue/comment that prompted the addition of the `UNMOUNTED_REDRAW` warning is here:

https://github.com/leeoniya/domvm/issues/131#issuecomment-299681306

this appears to have been fixed in 3.1.1 as a side-effect of mem leak patches for explicit views. updating the problematic fiddle with 3.2.3, the error no longer happens: https://jsfiddle.net/2qrgsk8b/8/

it's probably safe to ignore that notice in many cases, especially with the 3.2.3 patches.

as far as architecture goes, i tend to use declarative views more frequently and do global root redraw until i run into perf issues (which is rare). if your tabs are large and complex, you should take a look at toggling their visibility by diffing and patching just the attrs of their top-level nodes rather than regenerating the entire vtree simply to toggle `display: none`. There's an example of this here: http://leeoniya.github.io/domvm/demos/playground/#diff-patch
